### ISSUE TITLE
Remove dependency on "rbd" tool for mapping and unmapping volumes in RBD volume plugin

### ISSUE THREAD

**Issue by kokhang** [kubernetes/kubernetes#45190]
## Background:
Currently the RBD volume plugin shells out to `rbd` to map and unmap Ceph block volumes to the host. This creates a dependency on `rbd` and `ceph-common` to be installed on all the nodes.

In addition to this, RBD volume plugin shells out to `modprobe rbd` before mapping a volume. If this call results in an error, the mapping request fails.

Aside to `map` and `unmap`, the RBD plugin also shells out to `rbd lock` to lock block volumes and prevent them to be mounted in multiple hosts during a ReadWriteOnce.

## Proposal

- Use sysfs to map and unmap volume
- Relax modprobe result
- Use exclusive-lock to handle locking and fencing

### Map and unmap using sysfs 

For map and unmap, i propose using sysfs, which KRBD also uses. With sysfs, one could map and unmap rbd volumes by simply writing to sysfs RBD bus which are just files. For example, to map an RBD volume:

`$ echo "192.168.0.1 name=admin rbd foo" > /sys/bus/rbd/add`

When using sysfs, KRBD uses kernel keyring to store the secret, whose key is used to map the volume. This can be implemented using https://github.com/jsipprell/keyctl, which is a Golang native interface to the linux kernel keyring api.

### Not fail if `modprobe rbd` fails

Modprobe is a kernel operation that should only be done once to load the RBD module. I feel that we shouldn't be shelling out `modprobe rbd` every time a volume needs to be mounted. This is a one-time operation done by an admin who is manages the Kubernetes cluster. If `modprobe rbd` needs to be called, at least it shouldn't not fail the mapping if it fails. A warning would suffice.

The issue with calling `modprobe` is made evidently on self-hosted Kubernetes deployment. On self-hosted kubelet, one would have to add additional volume mounting to get it to work. ie mounting `/usr/bin/modprobe` and `/lib/modules` so that they are accessible in the kubelet self-hosted container.

### Exclusive-lock for fencing
Currently, the RBD volume plugin shells out to `rbd lock` to do volume locking. This is a legacy advisory type of locking and not enforced within librbd. 

What I propose is to use the `exclusive-lock` feature of ceph. https://www.sebastien-han.fr/blog/2015/07/06/ceph-enable-the-object-map-feature/.

The `exclusive-lock` is available in kernel 4.9+. From my conversation with @rootfs, there is already an issue that track this feature (https://github.com/kubernetes/kubernetes/issues/43809). There is also a fix proposed (https://github.com/rootfs/kubernetes/commit/3065efc77c7d7b743ca1bbd90d7b06c324760df4).

Since `exclusive-lock` is only available on kernel 4.9+, what we can do is check the kernel version and use exclusive-lock if it's supported or fall back to `rbd lock` if not supported.

Relying on `uname -r` might be ok. Yes, there could be cases where the kernel version dont change but patches that enable exclusive-locking are picked up and applied. At worst, this would just result in a false negative, and it would revert to the current implementation (rbd lock). 

## Summary
So in summary, my proposal to remove the dependency on `rdb` tool on Kubernetes host is to:

- Use sysfs to map and unmap rbd blocks
- Warn if `modprobe rbd` fails.
- Use exclusive-lock feature for locking/fencing volumes.
- Revert to `rbd lock` if exclusive-lock is not supported (< kernel v4.9)

**Comment by luxas** [comment]
cc @kubernetes/sig-storage-feature-requests @kubernetes/sig-storage-bugs

**Comment by rootfs** [comment]
@idryomov for sysfs interface.

**Comment by kokhang** [comment]
[Rook](www.github.com/rook/rook) is currently using sysfs to map/unmap. https://github.com/rook/rook/blob/master/cmd/rook/block/map.go#L124

**Comment by idryomov** [comment]
The sysfs interface is stable and we are committed to keeping it that way.  By omitting the tool you are missing out on the udev bits, but it's mostly just waiting until udev rules are processed before returning from `rbd map`, so probably not an issue.

One needs to write to `.../add_single_major` instead of `.../add` if the former is there:
```
/*
 * 'add' and 'add_single_major' interfaces are identical, but if rbd
 * kernel module is new enough and is configured to use single-major
 * scheme, 'add' is disabled in order to prevent old userspace from
 * doing weird things at unmap time.
 *
 * Same goes for 'remove' vs 'remove_single_major'.
 */
```

Unrelated, but the tool shouldn't fail if modprobe fails.  If it does, that's a bug.

Last, but _very_ important: exclusive-lock machinery cannot be reused for fencing with current kernels:  you need the new "exclusive" map option that disables automatic lock transfers (merged into 4.12-rc1).  
Revert to `rbd lock` needs to happen if the image cannot be mapped in exclusive mode.

**Comment by kokhang** [comment]
@rootfs Based on @idryomov's comment, this is what i envision the changes would be:
1- Use sysfs to map/unmap blocks instead of `rbd map`. Writing to `/add_single_major` and `/add` if single_major exists. This is something we already do in Rook (https://github.com/rook/rook/blob/master/cmd/rook/block/map.go#L89) and I could port this work to the rbd plugin.

2- Relax `modprobe rbd`. Not have it fail if it fails. Perhaps a warning.

3- Use exclusive-lock if kernel supports it. If it doenst, fall back to `rbd lock`. We could include your  PR (https://github.com/rootfs/kubernetes/commit/3065efc77c7d7b743ca1bbd90d7b06c324760df4). And we can add a check for kernel version to see whether to fall back to `rbd lock` or not.

Do you think this is something i can do for 1.7?

**Comment by rootfs** [comment]
@idryomov can sysfs disable lock owner transition?
@kokhang in addition to map, rbd is also needed for [provisioning](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/rbd/rbd_util.go#L344) and [checking status](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/rbd/rbd_util.go#L409)

**Comment by rootfs** [comment]
@kokhang how did you set key in sysfs? The [ceph secret](https://github.com/ceph/ceph/blob/master/src/common/secret.c#L74) looks using kernel key utils.

**Comment by idryomov** [comment]
@rootfs Yes.  exclusive is just another map option, like rw/ro or queue_depth.  You'd write:
```
<mon ip:port> name=<client id>,...,exclusive <pool name> <image name> -
```
instead of
```
<mon ip:port> name=<client id>,... <pool name> <image name> -
```
@kokhang Mapping an image with exclusive-lock feature enabled if the kernel supports exclusive-lock feature bit but doesn't support the new exclusive option (naming...!) should NOT be allowed by the plugin, because using `rbd lock` in such a case won't work.  Only two cases should be allowed:
- image with exclusive-lock feature disabled -- map w/o exclusive option and use `rbd lock`
- image with exclusive-lock feature enabled -- map with exclusive option.  If it fails with `bad option at ...`, bail and recommend disabling exclusive-lock feature on the image.

**Comment by kokhang** [comment]
@idryomov Since the new exclusive map option is available on 4.12, we could map an image with exclusive-lock if the kernel is 4.12+ and fall back to `rbd lock` otherwise.

@rootfs, for ceph secret, you can pass the secret to sysfs using the `secret=`, such as 
```$ echo "192.168.0.1 name=admin secret=<secret> rbd foo" > /sys/bus/rbd/add_single_major```

If you want to use the `key=`, there is a Golang Native implementation for kernel keyring. https://github.com/jsipprell/keyctl. That is something we can use.

As for using `rbd` to provision, this only have to be made available in the controller manager and not on all nodes. While it is not a complete solution, just removing the `rbd` dependency on the k8s hosts has merit. The controller-manager can be easily patched or worked around using containers for CM with ceph-tools installed etc. 

We could also leverage the `Attacher` interface (https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/volume.go#L161) so that RBD command is not used during the attach/detach on the nodes. And this will decouple the provisioning from the mapping in the RBD plugin.

**Comment by kfox1111** [comment]
For provisioning, I think it should be moved out to an external provisioner like the cephfs provisioner. Then you can start with a stock controller-manager container and add a rbd-provisioner container that matches your ceph version.

**Comment by idryomov** [comment]
@kokhang Kernels 4.9-4.11 support exclusive-lock feature bit, but not the new exclusive map option.  This is where the plugin should bail because the mapping will succeed, but `rbd lock` fallback won't work.

**Comment by kokhang** [comment]
@kfox1111 Agree. This is exactly what we do with Rook.

@rootfs, based on @idryomov's comment https://github.com/kubernetes/kubernetes/issues/45190#issuecomment-301898781,  here is my proposal for the locking mechanism:
For kernel < 4.12: Disable exclusive-lock feature on image creation, map without exclusive lock and use`rbd lock` 
For kernel >= 4.12, Enable exclusive-lock on image, map with exclusive lock.

Please let me know what you think as I'm willing to start this work ASAP.

**Comment by kfox1111** [comment]
Id recommend making the lock type overridable. a lot of vendor kernels report to be older then some of their subsystems are. With Redhat owning Ceph, I'd be surprised if their 3.10.x kernel didn't have newer ceph bits.

**Comment by kokhang** [comment]
@kfox1111 yes. As i specified in the proposal (first comment), relying on`uname -r` might be ok.  At worst, this would just result in a false negative.

**Comment by rootfs** [comment]
@kokhang 
- checking kernel version is not a reliable mechanism to check rbd features, e.g. centos backports 4.x kernel features into 3.x base. 
- can you check the if the rbd advisory lock is already held by other client via sysfs (currently done via `rbd status`)?

**Comment by kokhang** [comment]
@rootfs While it is true that checking the kernel version might not be reliable, how about if we go with that @kfox1111 suggested and make this as a config? We can add a flag in both the storageclass and the RBDVolumeSource to indicate whether to use exclusive-lock or not. By default we can set it to false, in which it will fall back on `rbd lock`, which is the current behavior. No need to do kernel version check with this option.

If we drive this based on a flag, during mapping, we dont have to check whether the image was created with exclusive-lock. We can just use the flag.

As for `rbd status`, we dont need to use sysfs to do this. If we get to this point, that means we are falling back to the advisory `rbd lock` and thus we can use `rbd status`.

So to be more exact, i am not going to totally get rid of `rbd`. We are still going to use it to fall back to advisory locking. We are also going to use it to provision images. This is just to get rid of `rbd` for map and unmap and use sysfs instead. This is just a small step to remove the dependency of RBD.

**Comment by rootfs** [comment]
If you wrap up map/unmap using sysfs and fall back to rbd cli, that sounds good to me.

**Comment by sbezverk** [comment]
@kokhang Do you have any plans to revive this initiative?
